Adds ExportOBJ Script

randomPoison opened the pull request:
- Adds option to file menu to export selected GameObject or prefab as an OBJ mesh.
- Also you can right click on the Transform component in the inspector to export the object.

randomPoison committed:
Adds ExportOBJ Script
- Adds option to file menu to export selected GameObject or prefab as an OBJ mesh.
- Also you can right click on the Transform component in the inspector to export the object.

brickman1444 commented on Assets/Editor/ExportOBJ.cs:
```diff
@@ -0,0 +1,158 @@
+
+	static void WriteToFile( string s, string filename )
+	{
+		using( StreamWriter sw = new StreamWriter( filename ) )
```
What is going on here?

cjacobwade replied:
Looks like it's putting the string s into a file. Using gets an object, makes it readonly for the duration of the block and then disposes of the reference: https://msdn.microsoft.com/en-us/library/yh598w02.aspx

brickman1444 commented:
Besides the one point where it uses C# I don't understand, everything looks fine. Open it up in Visual Studio and check for warnings.

brickman1444 commented on Assets/Editor/ExportOBJ.cs:
```diff
@@ -0,0 +1,158 @@
+	}
+}
+
+public class ObjExporter : ScriptableObject
```
Why is this a scriptable object? Is that the only way to add menu stuff?

randomPoison replied:
No, you can add menu items from any class as long as it's a static void method. I think scriptable object is needed to create files and write to them.

cjacobwade commented on Assets/Editor/ExportOBJ.cs:
```diff
@@ -0,0 +1,158 @@
+		{
+			Vector3 v = t.TransformPoint( vv );
+			numVertices++;
+			sb.Append( string.Format( "v {0} {1} {2}\n", v.x, v.y, -v.z ) );
```
David, didn't you say it was faster to use += when concatenating strings? Obviously this code isn't really performant heavy but string.Append is used a bunch in here and it doesn't match our standards.

brickman1444 replied:
This is using a StringBuilder. I'm not sure if that even has the += operator.

randomPoison replied:
[This article](http://www.dotnetperls.com/stringbuilder) talks about the performance of string builders down towards the bottom. For the most part, if you're going to do a lot of appending, then `StringBuilder.Append()` is going to be faster than +=. We can still speed this up, though, by telling the string builder from the beginning how big the final string is going to be. Since we know how many vertices and faces there are, we should be able to do that calculation before actually building the string.

cjacobwade commented on Assets/Editor/ExportOBJ.cs:
```diff
@@ -0,0 +1,158 @@
+		Debug.Log( "Exported Mesh: " + fileName );
+	}
+
+	static string processTransform( Transform t, bool makeSubmeshes )
```
Should be ProcessTransform

cjacobwade commented on Assets/Editor/ExportOBJ.cs:
```diff
@@ -0,0 +1,158 @@
+	public static void End()
+	{
+		StartIndex = 0;
+	}
```
Start and end are dumb method names to show the same thing happening...

randomPoison committed:
Script Cleanup
- Fixed unused variable warnings.
- Removed poorly names Start() and End() methods.
- Made the export menu option not show up under "File", and removed the "No Submeshes" option since it doesn't fit our use case.
